includes/gateway/class-omise-payment-billpayment-tesco.php

```diff
@@ -0,0 +1,232 @@
+			$metadata = array_merge(
+				apply_filters( 'omise_charge_params_metadata', array(), $order ),
+				array( 'order_id' => $order_id ) // override order_id as a reference for webhook handlers.
+			);
+
+			return OmiseCharge::create( array(
+				'amount'      => Omise_Money::to_subunit( $order->get_total(), $order->get_order_currency() ),
+				'currency'    => $order->get_order_currency(),
+				'description' => apply_filters( 'omise_charge_params_description', 'WooCommerce Order id ' . $order_id, $order ),
+				'source'      => array( 'type' => 'bill_payment_tesco_lotus' ),
+				'return_uri'  => add_query_arg( 'order_id', $order_id, site_url() . "?wc-api=omise_billpayment_tesco_callback" ),
+				'metadata'    => $metadata
+			) );
+		}
+
+		/**
+		 * @inheritdoc
+		 */
+		public function result( $order_id, $order, $charge ) {
+			if ( 'failed' == $charge['status'] ) {
+				return $this->payment_failed( $charge['failure_message'] . ' (code: ' . $charge['failure_code'] . ')' );
+			}
+
+			if ( 'pending' == $charge['status'] ) {
+				return [
+					'result'   => 'success',
+					'redirect' => $this->get_return_url( $order )
+				];
+			}
+
+			return $this->payment_failed(
+				sprintf(
+					__( 'Please feel free to try submitting your order again, or contact our support team if you have any questions (Your temporary order id is \'%s\')', 'omise' ),
+					$order_id
+				)
+			);
+		}
+
+		/**
+		 * @param int|WC_Order $order
+		 * @param string       $context  pass 'email' value through this argument only for 'sending out an email' case.
+		 */
+		public function display_barcode( $order, $context = 'view' ) {
+			if ( ! $this->load_order( $order ) ) {
+				return;
+			}
+
+			$charge_id    = $this->get_charge_id_from_order();
+			$charge       = OmiseCharge::retrieve( $charge_id );
+			$barcode_svg  = file_get_contents( $charge['source']['references']['barcode'] );
+			?>
+
+			<div class="omise omise-billpayment-tesco-details" <?php echo 'email' === $context ? 'style="margin-bottom: 4em; text-align:center;"' : ''; ?>>
+				<p><?php echo __( 'Use this barcode to pay at Tesco Lotus.', 'omise' ); ?></p>
+				<div class="omise-billpayment-tesco-barcode">
+					<?php
+					if ( 'email' === $context ) {
+						$barcode_html = $this->barcode_svg_to_html( $barcode_svg );
+						echo $barcode_html;
+					} else {
+						echo $barcode_svg;
+					}
+					?>
+				</div>
+				<small class="omise-billpayment-tesco-reference-number">
+					<?php
+					echo sprintf(
+						'| &nbsp; %1$s &nbsp; 00 &nbsp; %2$s &nbsp; %3$s &nbsp; %4$s',
+						$charge['source']['references']['omise_tax_id'],
+						$charge['source']['references']['reference_number_1'],
+						$charge['source']['references']['reference_number_2'],
+						$charge['amount']
+					);
+					?>
+				</small>
+
+				<?php if ( 'email' !== $context ) : ?>
+					<div class="omise-billpayment-tesco-print-button">
+						<button onClick="window.print()" class="button button-primary">Print barcode</button>
```

Comment:
This will print the entire page - not just the barcode. We should probably consider printing just the barcode ans some relevant info (as @jacstn already mentioned). This could probably be achieved by means of a print stylesheet temporarily applied, or perhaps creating an invisible iframe with some programmatically generated content in it, and printing that

Reply:
@jacstn @jonrandy thanks for the input.
I intended to leave it as it is because I want to let a stylesheet of a store's theme to handle the printing style in case if they want to have some product references on a paper when it get printed out. (i.e. product detail, price).

Having a simple print function is leaving a room for merchant/developer to remove it out later on if they want to (and user is still able to choose to print only a single page from a browser's printing dialog anyway) – on the other hand, forcing to print only barcode leaving no room for any modification (it could, but adding it back is technically more troublesome than removing it out). 

@jacstn @jonrandy ☝️ with the above reason, what do you guys think, still suggest to force the page to be printed out only barcode?

Reply:
not sure who wants to print anything more than a barcode, to pay to store,
but it is not a big issue, so I am not going to argue about that. AirAsia makes you print the only barcode with brief info what you should do with that.

@guzzilar how about this issue with making a layout that might introduce some problems if not customized by a merchant?

I mean.. this part with some extra icons..

Reply:
I still think we should really print just the barcode. If we are going to leave as it is now though, I would relabel the button 'Print this page' to give the user some warning they are about to kill more trees than is necessary

Reply:
If we programmatically create something in a hidden iframe, there is no issue with any conflicts with the merchant's checkout page design... but they would have no control over the way the barcode printout looked. Unless we could somehow pull a themeable layout from the server into said hidden iframe, fill in the details and barcode programmatically, then trigger a print there

Reply:
@jacstn @jonrandy 👍 thanks, ok, let's try with you guys suggested.
I'll add a hidden `<div></div>` element and set it to only display on `@media print` with some order informations.

After this, I have one PR that is allowing WordPress theme to override Omise template (#116). This should help merchant to be able to custom the printing detail later on.

Reply:
Ok, I said that let's try but after looked at a printing style of a theme I'm currently using:
![Screen Shot 2562-08-08 at 04 41 39](https://user-images.githubusercontent.com/2154669/62659978-09600a00-b997-11e9-8bc9-62e7a393e717.png)

![Screen Shot 2562-08-08 at 04 41 48](https://user-images.githubusercontent.com/2154669/62660001-14b33580-b997-11e9-96f9-89e280854772.png)

I think it's really a theme responsibility to proper style the printing page, not from plugin...
How about just change a button title to `print this page`?

👆 if you guys still vote for having a hidden div then I'll give up. But I'll pretty much just put everything that 90% similar to the above screenshots anyway... (payment detail, barcode, order detail).

Reply:
![Screen Shot 2562-08-08 at 04 48 16](https://user-images.githubusercontent.com/2154669/62660255-c6526680-b997-11e9-8f4b-3a2461722a1e.png)
btw, here is the actual screen of the above printing screen - screenshots.

Reply:
Ok, second thought after second thought (should it says "third thought"? lol). 
You guys are right, at least plugin should responsible for providing a default one with a proper way that 3rd-party could override and customize it at their desire.

Thanks for all the inputs 👍 👍